sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -835,5 +862,110 @@ public async ValueTask DisposeAsync()
+
+            lock (_maxConcurrencySyncLock)
+            {
+                var newConcurrency = _sessionIds.Length > 0
+                    ? Math.Min(_sessionIds.Length, maxConcurrentSessions)
+                    : maxConcurrentSessions * maxConcurrentCallsPerSession;
+
+                ReconcileConcurrency(maxConcurrentSessions, maxConcurrentCallsPerSession);
+
+                _maxConcurrentSessions = maxConcurrentSessions;
+                _maxConcurrentCallsPerSession = maxConcurrentCallsPerSession;
+                _maxConcurrentCalls = newConcurrency;
+            }
+        }
+
+        private void ReconcileConcurrency(int newConcurrency, int newMaxSessions = default)
+        {
+            // if processor isn't running there is nothing to reconcile
+            if (!IsProcessing)
+            {
+                return;
+            }
+
+            int diff = newConcurrency - _maxConcurrentCalls;
+            // increasing concurrency
+            if (diff > 0)
+            {
+                if (IsSessionProcessor)
+                {
+                    var diffSessions = newMaxSessions - _maxConcurrentSessions;
+                    if (diffSessions > 0 && _sessionIds.Length == 0)
+                    {
+                        for (int i = 0; i < diffSessions; i++)
+                        {
+                            _receiverManagers.Add(
+                                new SessionReceiverManager(
+                                    _sessionProcessor,
+                                    null,
+                                    MaxConcurrentAcceptSessionsSemaphore,
+                                    _scopeFactory,
+                                    _maxConcurrentCallsPerSession,
+                                    KeepOpenOnReceiveTimeout));
+                        }
+                    }
+                }
+                _messageHandlerSemaphore.Release(diff);
+            }
+            // decreasing concurrency
+            else if (diff < 0)
+            {
+                int diffLimit = Math.Abs(diff);
+                // limit the number of new tasks that can spawn to newly specified concurrency
+                for (int i = 0; i < diffLimit; i++)
+                {
+                    // use the async method to avoid blocking user callback thread
+                    _ = _messageHandlerSemaphore.WaitAsync();
+                }
+
+                var activeTasks = _tasks.Where(t => !t.Item1.IsCompleted).ToList();
+                int excessTasks = activeTasks.Count - newConcurrency;
+
+                // cancel excess tasks
+                for (int i = 0; i < excessTasks; i++)
+                {
+                    activeTasks[i].Item2.Cancel();
+                }
+
+                if (IsSessionProcessor)
+                {
+                    var diffSessions = _maxConcurrentSessions - newMaxSessions;
+                    if (diffSessions > 0 && _sessionIds.Length == 0)
```

Comment:
what if diffSessions< 0

Reply:
So if the total concurrency decreased, but the number of sessions increased, then we should be adding more managers. will fix.